stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,554 @@
+
+NEXT:
+	ta := a.time(i)
+	tb := b.time(j)
+
+	if ta == -1 || tb == -1 {
+		goto DONE
+	}
+
+	if ta < tb {
+		i++
+		goto NEXT
+	}
+
+	if ta > tb {
+		j++
+		goto NEXT
+	}
+
+	for k := j; ta == b.time(k); k++ {
+
+		// Evaluate fn over both input rows
+		obj, err := c.fn.Eval(c.ctx, a.record(i), b.record(k))
+		if err != nil {
+			return nil, err
+		}
+
+		// Build schema if this is the first row being joined
+		if firstRow {
+			if err := buildSchema(builder, obj); err != nil {
+				return nil, err
+			}
+			firstRow = false
+		}
+
+		// Check fn does not update the group key values.
+		// TODO(josh): This should be caught during planning.
+		// Remove this when the planner is made schema aware.
+		if ok := objContainsKey(obj, key); !ok {
+			return nil, errors.New(codes.Invalid, "argument 'fn' may not modify group key")
+		}
+
+		// The record obtained from calling fn may be added to output
+		if err := appendRowToBuilder(builder, obj); err != nil {
+			return nil, err
+		}
+	}
+	i++
+	goto NEXT
+DONE:
+	return builder.Table()
+}
+
+// objContainsKey checks if an object contains a specific group key
+func objContainsKey(obj values.Object, key flux.GroupKey) bool {
+	for _, col := range key.Cols() {
+		if v, ok := obj.Get(col.Label); !ok || !v.Equal(key.LabelValue(col.Label)) {
+			return false
+		}
+	}
+	return true
+}
+
+// buildSchema adds a schema defined by an object to an empty builder
+func buildSchema(builder *execute.ColListTableBuilder, obj values.Object) error {
+	schema := make([]flux.ColMeta, 0, obj.Len())
+	obj.Range(func(name string, v values.Value) {
+		schema = append(schema, flux.ColMeta{
+			Label: name,
+			Type:  execute.ConvertFromKind(v.Type().Nature()),
+		})
+	})
+	sort.Slice(schema, func(i, j int) bool {
+		return schema[i].Label < schema[j].Label
+	})
+	for _, col := range schema {
+		if _, err := builder.AddCol(col); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func appendRowToBuilder(builder *execute.ColListTableBuilder, obj values.Object) error {
+	var err error
+	obj.Range(func(name string, v values.Value) {
+		idx := execute.ColIdx(name, builder.Cols())
+		if idx < 0 {
+			err = errors.Newf(codes.NotFound, "column %s not found", name)
+			return
+		}
+		if err = builder.AppendValue(idx, v); err != nil {
+			return
+		}
+	})
+	return err
+}
+
+func NewRowIterator(readers []flux.ColReader, timeCol int) *RowIterator {
+	offsets, l := make([]int, len(readers)), 0
+	for i, r := range readers {
+		offsets[i] = l
+		l += r.Len()
+	}
+	return &RowIterator{
+		row:     make(map[string]values.Value),
+		len:     l,
+		readers: readers,
+		offsets: offsets,
+		timeCol: timeCol,
+	}
+}
+
+// RowIterator iterates over the rows of several column readers
+type RowIterator struct {
+	len int
+	row map[string]values.Value
+
+	timeCol int
+	offsets []int
+	readers []flux.ColReader
+}
+
+// columns returns the columns of the iterator
+func (iter *RowIterator) columns() []flux.ColMeta {
+	return iter.readers[0].Cols()
```

Comment:
Can you get these columns from the table instead and passed in explicitly? Indexing into the reader at index 0 makes me nervous. I'm not sure if there's a guarantee that `Do` will always return at least one table. There is certainly no test that mandates it.